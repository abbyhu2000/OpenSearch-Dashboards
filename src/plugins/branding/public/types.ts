import { NavigationPublicPluginStart } from '../../navigation/public';

export interface BrandingPluginSetup {
  getGreeting: () => string;
}
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface BrandingPluginStart {}

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
}
