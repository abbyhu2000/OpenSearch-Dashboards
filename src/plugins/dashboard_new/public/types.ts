/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { SavedObject } from '../../saved_objects/public';
import { Filter, ISearchSource, Query, RefreshInterval } from '../../data/public';
import { SavedDashboardPanel730ToLatest } from '../common';
import { ViewMode } from 'src/plugins/embeddable/public';

/**
 * This should always represent the latest dashboard panel shape, after all possible migrations.
 */
export type SavedDashboardPanel = SavedDashboardPanel730ToLatest;

export interface SavedDashboardState {
  panels: SavedDashboardPanel[];
  fullScreenMode: boolean;
  title: string;
  description: string;
  timeRestore: boolean;
  options: {
    hidePanelTitles: boolean;
    useMargins: boolean;
  };
  viewMode: ViewMode;
  expandedPanelId?: string;
}

export interface ISavedDashboard {
  id?: string;
  timeRestore: boolean;
  timeTo?: string;
  timeFrom?: string;
  description?: string;
  panelsJSON: string;
  optionsJSON?: string;
  // TODO: write a migration to rid of this, it's only around for bwc.
  uiStateJSON?: string;
  lastSavedTitle: string;
  refreshInterval?: RefreshInterval;
  searchSource?: ISearchSource;
  getQuery(): Query;
  getFilters(): Filter[];
}

export interface DashboardSavedObject extends SavedObject, ISavedDashboard {}

interface FieldSubType {
  multi?: { parent: string };
  nested?: { path: string };
}

export interface Field {
  name: string;
  type: string;
  // esTypes might be undefined on old index patterns that have not been refreshed since we added
  // this prop. It is also undefined on scripted fields.
  esTypes?: string[];
  aggregatable: boolean;
  filterable: boolean;
  searchable: boolean;
  subType?: FieldSubType;
}

export type NavAction = (anchorElement?: any) => void;

// TODO: this might not be needed because we have setDashboard(partial) now
// /**
//  * In URL panels are optional,
//  * Panels are not added to the URL when in "view" mode
//  */
// export type DashboardAppStateInUrl = Omit<DashboardAppState, 'panels'> & {
//   panels?: SavedDashboardPanel[];
// };

export interface SavedDashboardPanelMap {
  [key: string]: SavedDashboardPanel;
}

export interface StagedFilter {
  field: string;
  value: string;
  operator: string;
  index: string;
}
