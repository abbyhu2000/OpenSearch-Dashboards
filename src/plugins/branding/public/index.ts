import './index.scss';

import { BrandingPlugin } from './plugin';

// This exports static code and TypeScript types,
// as well as, OpenSearch Dashboards Platform `plugin()` initializer.
export function plugin() {
  return new BrandingPlugin();
}
export { BrandingPluginSetup, BrandingPluginStart } from './types';
