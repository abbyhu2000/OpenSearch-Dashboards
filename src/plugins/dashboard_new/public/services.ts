/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  ApplicationStart,
  Capabilities,
  ChromeStart,
  HttpStart,
  I18nStart,
  IUiSettingsClient,
  OverlayStart,
  SavedObjectsStart,
} from '../../../core/public';
import { TypesStart } from './vis_types';
import { createGetterSetter } from '../../opensearch_dashboards_utils/common';
import {
  DataPublicPluginStart,
  FilterManager,
  IndexPatternsContract,
  TimefilterContract,
} from '../../../plugins/data/public';
import { UsageCollectionSetup } from '../../usage_collection/public';
import { ExpressionsStart } from '../../expressions/public';
import { UiActionsStart } from '../../ui_actions/public';
import { SavedDashboardsLoader } from './saved_dashboards';
import { SavedObjectLoader } from '../../saved_objects/public';
import { EmbeddableStart } from '../../embeddable/public';

export const [getUISettings, setUISettings] = createGetterSetter<IUiSettingsClient>('UISettings');

export const [getCapabilities, setCapabilities] = createGetterSetter<Capabilities>('Capabilities');

export const [getHttp, setHttp] = createGetterSetter<HttpStart>('Http');

export const [getApplication, setApplication] = createGetterSetter<ApplicationStart>('Application');

export const [getEmbeddable, setEmbeddable] = createGetterSetter<EmbeddableStart>('Embeddable');

export const [getSavedObjects, setSavedObjects] = createGetterSetter<SavedObjectsStart>(
  'SavedObjects'
);

export const [getTypes, setTypes] = createGetterSetter<TypesStart>('Types');

export const [getI18n, setI18n] = createGetterSetter<I18nStart>('I18n');

export const [getFilterManager, setFilterManager] = createGetterSetter<FilterManager>(
  'FilterManager'
);

export const [getTimeFilter, setTimeFilter] = createGetterSetter<TimefilterContract>('TimeFilter');

export const [getIndexPatterns, setIndexPatterns] = createGetterSetter<IndexPatternsContract>(
  'IndexPatterns'
);

export const [getSearch, setSearch] = createGetterSetter<DataPublicPluginStart['search']>('Search');

export const [getUsageCollector, setUsageCollector] = createGetterSetter<UsageCollectionSetup>(
  'UsageCollection'
);

export const [getExpressions, setExpressions] = createGetterSetter<ExpressionsStart>('Expressions');

export const [getUiActions, setUiActions] = createGetterSetter<UiActionsStart>('UiActions');

export const [getSavedDashboardsLoader, setSavedDashboardsLoader] = createGetterSetter<
  SavedDashboardsLoader
>('SavedDashboardsLoader');

export const [getAggs, setAggs] = createGetterSetter<DataPublicPluginStart['search']['aggs']>(
  'AggConfigs'
);

export const [getOverlays, setOverlays] = createGetterSetter<OverlayStart>('Overlays');


