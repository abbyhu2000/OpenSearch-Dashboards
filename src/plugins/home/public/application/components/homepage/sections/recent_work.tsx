/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import useObservable from 'react-use/lib/useObservable';
import moment from 'moment';
import { EuiFlexItem, EuiCard, EuiIcon, EuiFlexGrid } from '@elastic/eui';
import { i18n } from '@osd/i18n';
import { ChromeRecentlyAccessedHistoryItem } from 'opensearch-dashboards/public';
import { Section } from '../../../../services/section_type/section_type';
import { renderFn } from './utils';
import { getServices } from '../../../opensearch_dashboards_services';

import '../_homepage.scss';

const itemType = [
  {
    type: 'dashboard',
    name: 'Dashboard',
    icon: 'dashboardApp',
  },
  {
    type: 'visualization',
    name: 'Visualization',
    icon: 'visualizeApp',
  },
  {
    type: 'search',
    name: 'Search',
    icon: 'discoverApp',
  },
  {
    type: 'visualization-visbuilder',
    name: 'Visualization(visbuilder)',
    icon: 'visualizeApp',
  },
];

const render = renderFn(() => {
  const services = getServices();
  const navigateToUrl = services.application.navigateToUrl;
  const recentAccessed = useObservable(services.chrome.recentlyAccessed.get$(), []);

  if (!recentAccessed.length) {
    return (
      <div className="empty-recent-work">
        <h2>No recent work</h2>
        <p>Recent work will appear here.</p>
      </div>
    );
  }

  return (
    <div>
      <EuiFlexGrid columns={4}>
        {recentAccessed.slice(0, 8).map((recentAccessItem: ChromeRecentlyAccessedHistoryItem) => {
          const recentWorkItem = itemType.filter((item) => item.type === recentAccessItem.type);
          if (recentWorkItem[0]) {
            return (
              <EuiFlexItem>
                <EuiCard
                  layout="horizontal"
                  title={i18n.translate('homepage.recentWorkSection.title', {
                    defaultMessage: recentAccessItem.label,
                  })}
                  titleSize="xs"
                  description={
                    <>
                      <EuiIcon
                        size="m"
                        className="recent-work-title-icon"
                        type={recentWorkItem[0].icon}
                      />
                      {i18n.translate('homepage.recentWorkSection.name', {
                        defaultMessage: recentWorkItem[0].name,
                      })}
                      <br />
                      {i18n.translate('homepage.recentWorkSection.updatedAt', {
                        defaultMessage:
                          'Last updated ' + moment(recentAccessItem?.updatedAt).fromNow(),
                      })}
                    </>
                  }
                  onClick={() => navigateToUrl(services.addBasePath(recentAccessItem.link))}
                />
              </EuiFlexItem>
            );
          }
        })}
      </EuiFlexGrid>
    </div>
  );
});

export const recentWorkSection: Section = {
  id: 'home:recentWork',
  title: i18n.translate('home.sections.recentWork.title', {
    defaultMessage: 'Recent work',
  }),
  render,
};
