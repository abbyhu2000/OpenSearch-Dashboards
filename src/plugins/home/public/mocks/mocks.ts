/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { applicationServiceMock, chromeServiceMock, coreMock } from '../../../../core/public/mocks';
import { HomeOpenSearchDashboardsServices } from '../application/opensearch_dashboards_services';
import { injectedMetadataServiceMock } from '../../../../core/public/mocks';
import { featureCatalogueRegistryMock } from '../services/feature_catalogue/feature_catalogue_registry.mock';
import { environmentServiceMock } from '../services/environment/environment.mock';
import { tutorialServiceMock } from '../services/tutorials/tutorial_service.mock';
import { sectionTypeServiceMock } from '../services/section_type/section_type.mock';
import { configSchema } from '../../config';

const createSetupContract = () => ({
  featureCatalogue: featureCatalogueRegistryMock.createSetup(),
  environment: environmentServiceMock.createSetup(),
  tutorials: tutorialServiceMock.createSetup(),
  sectionTypes: sectionTypeServiceMock.createSetup(),
});

export const createStartContract = () => {
  const coreStartMock = coreMock.createStart();
  const chromeMock = chromeServiceMock.createStartContract();
  const applicationMock = applicationServiceMock.createStartContract();
  const toastNotifications = coreStartMock.notifications.toasts;
  const injectedMetadataMock = injectedMetadataServiceMock.createStartContract();
  injectedMetadataMock.getBranding.mockReturnValue({});
  const addBasePathMock = jest.fn((path: string) => (path ? path : 'path'));
  const opensearchDashboardsVersion = '3.0.0';

  return ({
    ...coreStartMock,
    chrome: chromeMock,
    application: applicationMock,
    injectedMetadata: injectedMetadataMock,
    addBasePath: addBasePathMock,
    getBasePath: jest.fn(),
    toastNotifications,
    opensearchDashboardsVersion,
    trackUiMetric: jest.fn(),
    savedObjectsClient: {
      find: async () => {
        return {
          total: 0,
        };
      },
    },
  } as unknown) as jest.Mocked<HomeOpenSearchDashboardsServices>;
};

export const homePluginMock = {
  createSetupContract,
  createStartContract,
};
