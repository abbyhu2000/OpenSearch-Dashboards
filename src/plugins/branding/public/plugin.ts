import { i18n } from '@osd/i18n';
import { PluginInitializerContext } from 'opensearch-dashboards/server';
import { AppMountParameters, CoreSetup, CoreStart, Plugin } from '../../../core/public';
import { BrandingPluginSetup, BrandingPluginStart, AppPluginStartDependencies } from './types';
import { PLUGIN_NAME } from '../common';
import { ConfigSchema } from '../config';
import { BrandingAssignment } from './services';

export class BrandingPlugin implements Plugin<BrandingPluginSetup, BrandingPluginStart> {
  private readonly brandingAssignment = new BrandingAssignment();

  constructor(private readonly initializerContext: PluginInitializerContext<ConfigSchema>) {}

  public setup(core: CoreSetup): BrandingPluginSetup {
    // Register an application into the side navigation menu
    core.application.register({
      id: 'branding',
      title: PLUGIN_NAME,
      async mount(params: AppMountParameters) {
        // Load application bundle
        const { renderApp } = await import('./application');
        // Get start services as specified in opensearch_dashboards.json
        const [coreStart, depsStart] = await core.getStartServices();
        // Render the application
        return renderApp(coreStart, depsStart as AppPluginStartDependencies, params);
      },
    });

    // Return methods that should be available to other plugins
    return {
      getGreeting() {
        return i18n.translate('branding.greetingText', {
          defaultMessage: 'Hello from {name}!',
          values: {
            name: PLUGIN_NAME,
          },
        });
      },
    };
  }

  public start(core: CoreStart): BrandingPluginStart {
    return {};
  }

  public stop() {}
}
