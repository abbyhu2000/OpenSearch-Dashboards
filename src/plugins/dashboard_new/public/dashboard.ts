/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/**
 * @name Dashboard
 *
 * @description This class consists of aggs, params, listeners, title, and type.
 *  - Aggs: Instances of IAggConfig.
 *  - Params: The settings in the Options tab.
 *
 * Not to be confused with vislib/vis.js.
 */

import { isFunction, defaults, cloneDeep } from 'lodash';
import { Assign } from '@osd/utility-types';
import { i18n } from '@osd/i18n';
import { PersistedState } from './persisted_state';
import { getTypes, getAggs, getSearch, getSavedSearchLoader } from './services';
import {
  IAggConfigs,
  IndexPattern,
  ISearchSource,
  AggConfigOptions,
  SearchSourceFields,
} from '../../../plugins/data/public';
import { SerializedDashboard } from '../dashboard';

export interface DashboardParams {
  [key: string]: any;
}

// type PartialDashboardState = Assign<SerializedVis, { data: Partial<SerializedVisData> }>;

// id?: string;
// timeRestore: boolean;
// timeTo?: string;
// timeFrom?: string;
// description?: string;
// panelsJSON: string;
// optionsJSON?: string;
// // TODO: write a migration to rid of this, it's only around for bwc.
// uiStateJSON?: string;
// lastSavedTitle: string;
// refreshInterval?: RefreshInterval;
// searchSource: ISearchSource;
// getQuery(): Query;
// getFilters(): Filter[];

export class Dashboard<TDashboardParams = DashboardParams> {
  public readonly id?: string;
  public title: string = '';
  public description: string = '';
  public params: TDashboardParams;
  // Session state is for storing information that is transitory, and will not be saved with the visualization.
  // For instance, map bounds, which depends on the view port, browser window size, etc.
  public sessionState: Record<string, any> = {};

  public readonly uiState: PersistedState;

  constructor(dashboardState: SerializedDashboard = {} as any) {
    this.params = this.getParams(dashboardState.params);
    this.uiState = new PersistedState(dashboardState.uiState);
    this.id = dashboardState.id;
  }

  private getParams(params: VisParams) {
    return defaults({}, cloneDeep(params ?? {}), cloneDeep(this.type.visConfig?.defaults ?? {}));
  }

  async setState(state: PartialDashboardState) {
    let typeChanged = false;
    if (state.type && this.type.name !== state.type) {
      // @ts-ignore
      this.type = this.getType(state.type);
      typeChanged = true;
    }
    if (state.title !== undefined) {
      this.title = state.title;
    }
    if (state.description !== undefined) {
      this.description = state.description;
    }
    if (state.params || typeChanged) {
      this.params = this.getParams(state.params);
    }
    if (state.data && state.data.searchSource) {
      this.data.searchSource = await getSearch().searchSource.create(state.data.searchSource!);
      this.data.indexPattern = this.data.searchSource.getField('index');
    }
    if (state.data && state.data.savedSearchId) {
      this.data.savedSearchId = state.data.savedSearchId;
      if (this.data.searchSource) {
        this.data.searchSource = await getSearchSource(
          this.data.searchSource,
          this.data.savedSearchId
        );
        this.data.indexPattern = this.data.searchSource.getField('index');
      }
    }
    if (state.data && (state.data.aggs || !this.data.aggs)) {
      const aggs = state.data.aggs ? cloneDeep(state.data.aggs) : [];
      const configStates = this.initializeDefaultsFromSchemas(aggs, this.type.schemas.all || []);
      if (!this.data.indexPattern) {
        if (aggs.length) {
          const errorMessage = i18n.translate(
            'visualizations.initializeWithoutIndexPatternErrorMessage',
            {
              defaultMessage: 'Trying to initialize aggs without index pattern',
            }
          );
          throw new Error(errorMessage);
        }
        return;
      }
      this.data.aggs = getAggs().createAggConfigs(this.data.indexPattern, configStates);
    }
  }

  clone() {
    const { data, ...restOfSerialized } = this.serialize();
    const vis = new Vis(this.type.name, restOfSerialized as any);
    vis.setState({ ...restOfSerialized, data: {} });
    const aggs = this.data.indexPattern
      ? getAggs().createAggConfigs(this.data.indexPattern, data.aggs)
      : undefined;
    vis.data = {
      ...this.data,
      aggs,
    };
    return vis;
  }

  serialize(): SerializedVis {
    const aggs = this.data.aggs ? this.data.aggs.aggs.map((agg) => agg.toJSON()) : [];
    return {
      id: this.id,
      title: this.title,
      description: this.description,
      type: this.type.name,
      params: cloneDeep(this.params) as any,
      uiState: this.uiState.toJSON(),
      data: {
        aggs: aggs as any,
        searchSource: this.data.searchSource ? this.data.searchSource.getSerializedFields() : {},
        savedSearchId: this.data.savedSearchId,
      },
    };
  }

  // deprecated
  isHierarchical() {
    if (isFunction(this.type.hierarchicalData)) {
      return !!this.type.hierarchicalData(this);
    } else {
      return !!this.type.hierarchicalData;
    }
  }

  private initializeDefaultsFromSchemas(configStates: AggConfigOptions[], schemas: any) {
    // Set the defaults for any schema which has them. If the defaults
    // for some reason has more then the max only set the max number
    // of defaults (not sure why a someone define more...
    // but whatever). Also if a schema.name is already set then don't
    // set anything.
    const newConfigs = [...configStates];
    schemas
      .filter((schema: any) => Array.isArray(schema.defaults) && schema.defaults.length > 0)
      .filter(
        (schema: any) => !configStates.find((agg) => agg.schema && agg.schema === schema.name)
      )
      .forEach((schema: any) => {
        const defaultSchemaConfig = schema.defaults.slice(0, schema.max);
        defaultSchemaConfig.forEach((d: any) => newConfigs.push(d));
      });
    return newConfigs;
  }
}
