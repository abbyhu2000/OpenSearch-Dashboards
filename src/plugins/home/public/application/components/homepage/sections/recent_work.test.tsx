/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { recentWorkSection } from './recent_work';
import {
  HomeOpenSearchDashboardsServices,
  setServices,
} from '../../../opensearch_dashboards_services';
import { createHomeServicesMock } from 'src/plugins/home/public/mocks';
import { BehaviorSubject } from 'rxjs';
import { ChromeRecentlyAccessedHistoryItem } from 'opensearch-dashboards/public';

let mockElement: HTMLElement;
let services: HomeOpenSearchDashboardsServices;

describe('Recent work section', () => {
  beforeAll(() => {
    mockElement = document.createElement('element');
    services = createHomeServicesMock();
  });

  it('renders empty recent work', async () => {
    services.chrome.recentlyAccessed.get$ = () => {
      return new BehaviorSubject<ChromeRecentlyAccessedHistoryItem[]>([]);
    };
    setServices(services);
    recentWorkSection.render(mockElement);
    expect(mockElement).toMatchSnapshot();
  });

  it('renders non-empty recent work sections with visualization and dashboard saved object', async () => {
    services.chrome.recentlyAccessed.get$ = () => {
      return new BehaviorSubject<ChromeRecentlyAccessedHistoryItem[]>([
        {
          link: 'link1',
          label: 'label1',
          id: 'id1',
          type: 'visualization',
        },
        {
          link: 'link2',
          label: 'label2',
          id: 'id2',
          type: 'dashboard',
        },
      ]);
    };
    setServices(services);
    recentWorkSection.render(mockElement);
    expect(mockElement).toMatchSnapshot();
  });
});
