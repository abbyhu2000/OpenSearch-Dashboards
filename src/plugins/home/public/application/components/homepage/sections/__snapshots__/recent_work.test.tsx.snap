// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recent work section renders empty recent work 1`] = `
<element>
  <div
    class="empty-recent-work"
  >
    <h2>
      No recent work
    </h2>
    <p>
      Recent work will appear here.
    </p>
  </div>
</element>
`;

exports[`Recent work section renders non-empty recent work sections with visualization and dashboard saved object 1`] = `
<element>
  <div>
    <div
      class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--fourths euiFlexGrid--responsive"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--isClickable euiCard euiCard--centerAligned euiCard--horizontal euiCard--isClickable"
        >
          <div
            class="euiCard__content"
          >
            <span
              class="euiTitle euiTitle--xsmall euiCard__title"
              id="generated-idTitle"
            >
              <button
                aria-describedby=" generated-idDescription"
                class="euiCard__titleButton"
              >
                label1
              </button>
            </span>
            <div
              class="euiText euiText--small euiCard__description"
              id="generated-idDescription"
            >
              <p>
                <span
                  class="recent-work-title-icon"
                  data-euiicon-type="visualizeApp"
                />
                Visualization
                <br />
                Last updated a few seconds ago
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem"
      >
        <div
          class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--isClickable euiCard euiCard--centerAligned euiCard--horizontal euiCard--isClickable"
        >
          <div
            class="euiCard__content"
          >
            <span
              class="euiTitle euiTitle--xsmall euiCard__title"
              id="generated-idTitle"
            >
              <button
                aria-describedby=" generated-idDescription"
                class="euiCard__titleButton"
              >
                label2
              </button>
            </span>
            <div
              class="euiText euiText--small euiCard__description"
              id="generated-idDescription"
            >
              <p>
                <span
                  class="recent-work-title-icon"
                  data-euiicon-type="dashboardApp"
                />
                Dashboard
                <br />
                Last updated a few seconds ago
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</element>
`;
